### PR TITLE
fix(index): use `emit` hook instead of `this-compilation`

### LINKED ISSUE #7: CopyWebpackPlugin assets are not getting minified
I believe this is due to the event listened on. It should rather be `emit` and not `this-compilation`, which makes sense, I  believe. Compressing should always be the last step and not happen between two output modifications as the minified content is outdated then (or, as in this case, not catched).

### LINKED ISSUE #9: Does not compress html files from html-webpack-plugin
Probably the same as in [compression-webpack-plugin](https://github.com/webpack-contrib/compression-webpack-plugin/issues/30), compressing html files from a previous generation step does not work. Can we do the same [fix](https://github.com/webpack-contrib/compression-webpack-plugin/pull/71) here?

### PR DESCRIPTION
**Pull request by eseliger**
### `Issues`

- Fixes #7 
- Fixes #9

### `Notable Changes`

Made the change, as discussed in #7

### CODE REVIEW

**Comment by eseliger** [comment]
the build is failing unrelated to the changes (nsp warning)

File: src/index.js [hunk 1]
```diff
@@ -32,55 +32,54 @@ class ZopfliPlugin {
...
-    compiler.plugin('this-compilation', (compilation) => {
-      compilation.plugin('optimize-assets', (assets, callback) => {
-        async.forEach(Object.keys(assets), (file, cb) => {
-          if (Array.isArray(this.test)) {
-            if (this.test.every(t => !t.test(file))) {
-              return cb();
-            }
-          } else if (this.test && !this.test.test(file)) {
+    compiler.plugin('emit', (compilation, callback) => {
+      const { assets } = compilation;
```
Comment (michael-ciniawsky) [inline src/index.js:36]: `\n`

File: src/index.js [hunk 2]
```diff
@@ -32,55 +32,54 @@ class ZopfliPlugin {
...
+      async.forEach(Object.keys(assets), (file, cb) => {
+        if (Array.isArray(this.test)) {
+          if (this.test.every(t => !t.test(file))) {
             return cb();
           }
-          const asset = assets[file];
-          let content = asset.source();
+        } else if (this.test && !this.test.test(file)) {
+          return cb();
+        }
```
Comment (michael-ciniawsky) [inline src/index.js:45]: `\n`

File: src/index.js [hunk 3]
```diff
@@ -32,55 +32,54 @@ class ZopfliPlugin {
...
-            return cb();
-          }
+        if (originalSize < this.threshold) {
+          return cb();
+        }
 
-          this.algorithm(content, this.compressionOptions, (err, result) => {
-            if (err) { return cb(err); }
+        this.algorithm(content, this.compressionOptions, (err, result) => {
+          if (err) { return cb(err); }
```
Comment (michael-ciniawsky) [inline src/index.js]: Please don't 'inline' functions body's, whitespace is cheap :)

```diff
-  if (err) { return cb(err); }
+ if (err) {
+  return cb(err);
+ }
```

File: src/index.js [hunk 4]
```diff
@@ -32,55 +32,54 @@ class ZopfliPlugin {
...
-            };
+          const parse = url.parse(file);
+          const sub = {
+            file,
+            path: parse.pathname,
+            query: parse.query || '',
+          };
 
-            let newFile = this.asset.replace(/\[(file|path|query)\]/g, (p0, p1) => sub[p1]);
+          let newFile = this.asset.replace(/\[(file|path|query)\]/g, (p0, p1) => sub[p1]);
```
Comment (michael-ciniawsky) [inline src/index.js]: `newFile` => `newAsset`

File: src/index.js [hunk 5]
```diff
@@ -32,55 +32,54 @@ class ZopfliPlugin {
...
 
-            if (typeof this.filename === 'function') {
-              newFile = this.filename(newFile);
-            }
-            assets[newFile] = new RawSource(result);
+          if (typeof this.filename === 'function') {
+            newFile = this.filename(newFile);
+          }
```
Comment (michael-ciniawsky) [inline src/index.js]: `\n`

File: src/index.js (hunk 4 above)
Comment (michael-ciniawsky) [inline src/index.js]: `\n`

**Review by michael-ciniawsky — APPROVED** [review]
@eseliger Thx

**Comment by eseliger** [comment]
@michael-ciniawsky I can change the style, if you want, but I didn't change it, i just indented these rows, thats why they are in the diff. It was like that before. 
Let me know what you think

**Comment by michael-ciniawsky** [comment]
Updating them would be appreciated :)

**Review by alexander-akait — APPROVED** [review]

**Comment by eseliger** [comment]
@michael-ciniawsky updated the style, let me know if I missed something! :)

**Comment by eseliger** [comment]
is there anything I can do to bring this forward? :)

**Review by michael-ciniawsky — APPROVED** [review]
@eseliger Thx

**Comment by eseliger** [comment]
All the time! Let me know when you'll release it on npm

**Comment by michael-ciniawsky** [comment]
Unfortunately I realised that I don't have `npm` publish access to this repo (yet) atm. Which hopefully will be fixed within the next 1-2 days or someone else cuts a release (I pinged folks already :)). We will release it asap whenever possible

**Comment by eseliger** [comment]
Okay, awesome!

**Comment by Kaesebrot84** [comment]
@michael-ciniawsky Any news?

**Comment by fabioscsilva** [comment]
@michael-ciniawsky @eseliger Any update? This is a great improvement that we are waiting for.

**Pull request merged**